### PR TITLE
Csrf edit

### PR DESCRIPTION
**Pull request by geogramdotcom**
Pull request for CSRF update

### CODE REVIEW

File: source/_posts/angular-csrf-protection-implementation.md [hunk 1]
```diff
@@ -149,45 +149,264 @@ The secure flag prevents this attack by the browser guaranteeing that the cookie
...
- secure: true,
- maxAge: 3600,
- httpOnly: true,
- sameSite: strict
-});
- var parseForm = bodyParser.urlencoded({ extended: false })
+/// Use Express
+var express = require("express");
+var session = require("express-session")
+// Use body-parser
```
Comment (bartosz-io) [inline source/_posts/angular-csrf-protection-implementation.md:185]: What is the value of these comments? They do not seem to add anything...

File: source/_posts/angular-csrf-protection-implementation.md [hunk 2]
```diff
@@ -149,45 +149,264 @@ The secure flag prevents this attack by the browser guaranteeing that the cookie
...
 
-The following is an example of the implementation of these methods in NodeJS. The user makes a GET request to the '/form' API endpoint and then when the user fills out the form, using angular form control and form builder, submits the data to the server via a POST request at the '/process' API endpoint.
+##### CSURF Package Vulnerability
 
+For those who have implimented a CSRF prevention method using CSURF, a popular Node.js CSRF protection middleware, there has been a rediscovered vulnerability. The `cookie: true` flag set is the trouble, and in short, an attacker can use cookie tossing (setting a cookie from a subdomain) to use a valid (and signed) pair of cookies (_csrf + token) to bypass the anti-csrf mechanism. This vulnerability is not present when `cookie: false` is set; however in abundance of caution the package author has depricated this module. 
+
+Instead of using the CSURF npm package, you should use the following implimentation on the server.
```
Comment (bartosz-io) [inline source/_posts/angular-csrf-protection-implementation.md:156]: implimentation -> typo

**Comment by bartosz-io** [comment]
How about adding a quote block with a">" sign at **the very beginning** of the article:

> The article was updated on 20th December 2022 to reflect the current state of XYZ and ABC.

Or something similar?

File: source/_posts/angular-csrf-protection-implementation.md [hunk 3]
```diff
@@ -133,7 +133,7 @@ Nothing is worse than going through all the effort to secure your website with a
 
 ##### Cookie Security; HttpOnly flag
 
-To block the potential liability of Javascript accessing the cookie and user data there is a flag that can be implemented to prevent this type of attack. The “HttpOnly” flag is used to block Javascript from accessing cookies from the client-side so if an attacker was to succeed in injecting some javascript.
+To block the potential liability of Javascript accessing the cookie and user data there is a flag that can be implemented to prevent this type of attack. The “HttpOnly” flag is used to block Javascript from accessing cookies from the client-side so if an attacker was to succeed in injecting some javascript it wouldn't be able to read any of the cookies.
```
Comment (bartosz-io) [inline source/_posts/angular-csrf-protection-implementation.md:136]: HttpOnly => `HttpOnly` (with ticks \`\`)

File: source/_posts/angular-csrf-protection-implementation.md [hunk 4]
```diff
@@ -149,45 +149,264 @@ The secure flag prevents this attack by the browser guaranteeing that the cookie
...
+
+Instead of using the CSURF npm package, you should use the following implimentation on the server. 
+
+## Implementation of CSRF Protection
+
+There is a big difference between the proper implementations of CSRF protection on stateful servers, where you have access to memory to store the state, and stateless services where you might use container instances or a serverless architecture. 
+
+For implementation of CSRF Protection on a stateful system, you will need to use the Synchronizer Token Pattern. The CSRF Token is generated server-side for this pattern along with a session id. **For the Synchronised Token Pattern, CSRF tokens should not be transmitted using cookies.** 
+
+When the client receives the CSRF token from the server, there are two options to use for the server to validate that the request is legitimate.
```
Comment (bartosz-io) [inline source/_posts/angular-csrf-protection-implementation.md:164]: A new line here?

**Comment by bartosz-io** [comment]
How about adding some simple visuals (diagram or graph) explaining CSRF ?

You could take the ones from the Web Security Academy lesson about CSRF :)

- [ ] Consider adding a visual

**Comment by bartosz-io** [comment]
@geogramdotcom any updated on this?

File: source/_posts/angular-csrf-protection-implementation.md [hunk 5]
```diff
@@ -57,7 +57,7 @@ The best way to stop CSRF attacks for Angular apps is using the default synchron
 
 To enable Angular CSRF protection in an application three things need to happen:
 
-1) Every XHR request sent out has to include a custom HTTP header with a specific name and value; this can also be coupled with the user's session id token called "double submit cookie method". Be wary of this implementation however since it is known to have vulnerabilities when you don't control all the sub-domains of the origin.
+1) Every XHR request sent out has to include a custom HTTP header with a specific name and value; this can also be coupled with the user's session id token called "double submit cookie method". Be wary of this implementation however since it is known to have vulnerabilities when you don't control all the sub-domains of the origin. This Double Submit Cookie Method should also only be used by a stateless application where, for example, you might be hosting your server as a container on Google Cloud Run or Amazon AWS Elastic instances. (See below for stateful implimentation of CSRF Protection)
```
Comment (devcodedcom) [inline source/_posts/angular-csrf-protection-implementation.md:60]: TYPO in `implimentation `

**Review by devcodedcom — CHANGES_REQUESTED** [review]

File: source/_posts/angular-csrf-protection-implementation.md [hunk 6]
```diff
@@ -15,6 +15,8 @@ relatedPost: angular-xss
 ## Table of Contents
 <!-- toc -->
 
+**UPDATE:** This article was updated Feb. 25, 2023 to include information on CSURF Package Vulnerability. For those who have implimented a CSRF prevention method using CSURF, a popular Node.js CSRF protection middleware, please read on to understand this vulnerability and what to do to avoid it.
```
[the remark this answers is not in the document]
Comment (devcodedcom) [reply source/_posts/angular-csrf-protection-implementation.md:18]: Also TYPO in implimented ->implemented

**Review by devcodedcom — CHANGES_REQUESTED** [review]

File: source/_posts/angular-csrf-protection-implementation.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (geogramdotcom) [reply source/_posts/angular-csrf-protection-implementation.md:18]: Removed the Date, but left the update note.

File: source/_posts/angular-csrf-protection-implementation.md [hunk 7]
```diff
@@ -149,45 +149,264 @@ The secure flag prevents this attack by the browser guaranteeing that the cookie
...
+// create application/json parser
+var jsonParser = bodyParser.json()
+
+// These settings are only for local development testing.
+// Do not use these in production.
+// In production, ensure you're using cors and helmet and have proper configuration.
+const { invalidCsrfTokenError, generateToken, doubleCsrfProtection } =
+  doubleCsrf({
+    getSecret: (req) => req.secret,
+    secret: CSRF_SECRET,
```
[the remark this answers is not in the document]
Comment (geogramdotcom) [reply source/_posts/angular-csrf-protection-implementation.md:368]: `secret` is correct per csrf-csrf docs https://github.com/Psifi-Solutions/csrf-csrf/blob/9be23b3bd28a69caec99c5951a5a0ac7968168cc/README.md?plain=1#L195

**Review by devcodedcom — APPROVED** [review]

**Review by bartosz-io — APPROVED** [review]

**Comment by geogramdotcom** [comment]
👏

**Pull request merged**